trie/sync.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"gopkg.in/karalabe/cookiejar.v2/collections/prque"
+)
+
+// request represents a scheduled or already in-flight state retrieval request.
```

Comment:
Data retrievals in the `downloader/queue` is a simple flat list of hashes: "gimme this; here ya' go". The `request`s here form a trie themselves, and reassemble the trie in memory, writing out pieces as they get completed. The `downloader/queue` makes sure that required data is retrieved, handles concurrent retrievals from multiple peers, timeouts, retries, ignore lists, etc. This piece of code maintains a prioritized ordering of "future" requests that should be executed, an order that changes at every state insertion. The `downloader/queue` cannot reprioritize its fetch order as it has no idea which stte data is more important than the other... you need to interpret the data for that. So the downloader queue maintains only a very small list of currently active state downloads to handle any retrieval failures and retries, and the sync.trie will maintain the real queue that always returns the most urgent nodes to fetch.